File: tests/lexer.js
import TokenType from './tokenTypes.js';

class Lexer {
    constructor(input) {
        this.pos = 0;
        this.input = input;
        this.restrictedKeywords = new Set([
            'var', 'const', 'for', 'switch', 'case', 'break', 
            'continue', 'default', 'class', 'extends', 'super', 'this', 
            'typeof', 'instanceof', 'delete', 'new', 'in', 
            'try', 'catch', 'finally', 'throw', 'debugger',
        ]);
    }

    nextToken() {
        while (this.pos < this.input.length) {
            const currentChar = this.input[this.pos];
    
            // Skip whitespace
            if (/\s/.test(currentChar)) {
                this.pos++;
                continue;
            }
    
            // Check for single-line comments
            if (currentChar === '/') {
                if (this.input[this.pos + 1] === '/') {
                    while (this.pos < this.input.length && this.input[this.pos] !== '\n') {
                        this.pos++;
                    }
                    this.pos++; // Move past the newline
                    continue;
                }
            }
    
            // Check for boolean literals
            if (this.input.startsWith('true', this.pos)) {
                this.pos += 4; // Move position past "true"
                return { value: true, type: TokenType.Boolean };
            }
            if (this.input.startsWith('false', this.pos)) {
                this.pos += 5; // Move position past "false"
                return { value: false, type: TokenType.Boolean };
            }
    
            // Check for numbers, strings, keywords, operators, and punctuation
            if (/\d/.test(currentChar)) {
                return this.readNumber();
            }
            if (currentChar === '"' || currentChar === "'" || currentChar === "`") {
                return this.readString(currentChar);
            }
            if (/[a-zA-Z_]/.test(currentChar)) {
                return this.readKeywordOrIdentifier();
            }
    
            // Handle operators and punctuation
            if (currentChar === '=') {
                this.pos++;
                if (this.input[this.pos] === '=') {
                    this.pos++;
                    return { value: '==', type: TokenType.Equals }; // Handle equality
                }
                return { value: '=', type: TokenType.Equals }; // Assignment
            }
            if (currentChar === '(') {
                this.pos++;
                return { value: '(', type: TokenType.OpenParen };
            }
            if (currentChar === ')') {
                this.pos++;
                return { value: ')', type: TokenType.CloseParen };
            }
            // Ignore semicolon
            if (currentChar === ';') {
                this.pos++;
                continue; // Skip semicolon
            }
            if (currentChar === '{') {
                this.pos++;
                return { value: '{', type: TokenType.OpenBrace };
            }
            if (currentChar === '}') {
                this.pos++;
                return { value: '}', type: TokenType.CloseBrace };
            }
    
            // Handle array syntax
            if (currentChar === '[') {
                this.pos++;
                return { value: '[', type: TokenType.OpenBracket };
            }
            if (currentChar === ']') {
                this.pos++;
                return { value: ']', type: TokenType.CloseBracket };
            }
            if (currentChar === ',') {
                this.pos++;
                return { value: ',', type: TokenType.Comma };
            }

            if (currentChar === '*' && this.input[this.pos + 1] === '*') {
                this.pos += 2; // Move past '**'
                return { value: '**', type: TokenType.BinaryOperator };
            }            
    
            // Handle binary operators
            if (['+', '-', '*', '/', '%', '^'].includes(currentChar)) {
                return this.readBinaryOperator(currentChar);
            }
    
            // Handle comparison operators
            if (currentChar === '>') {
                return this.readComparisonOperator('>');
            }
            if (currentChar === '<') {
                return this.readComparisonOperator('<');
            }

            // Handle logical operators
            if (currentChar === '&' && this.input[this.pos + 1] === '&') {
                this.pos += 2; // Move past '&&'
                return { value: '&&', type: TokenType.LogicalAnd };
            }
            if (currentChar === '|' && this.input[this.pos + 1] === '|') {
                this.pos += 2; // Move past '||'
                return { value: '||', type: TokenType.LogicalOr };
            }

            // Handle not equal operator
            if (currentChar === '!' && this.input[this.pos + 1] === '=') {
                this.pos += 2; // Move past '!='
                return { value: '!=', type: TokenType.NotEquals };
            }
    
            throw new Error(`Unexpected character: ${currentChar}`);
        }
    
        return null; 
    }

    readKeywordOrIdentifier() {
        let idStr = '';
        
        // Allow the first character to be a letter or underscore
        if (/[a-zA-Z_]/.test(this.input[this.pos])) {
            idStr += this.input[this.pos++];
        } else {
            throw new Error(`Invalid identifier start: ${this.input[this.pos - 1]} at position ${this.pos - 1}`);
        }
    
        // Allow subsequent characters to be letters, digits, or underscores
        while (this.pos < this.input.length && /[a-zA-Z0-9_]/.test(this.input[this.pos])) {
            idStr += this.input[this.pos++];
        }
    
        // Check for restricted keywords
        if (this.restrictedKeywords.has(idStr) && idStr !== 'delvar all') {
            throw new Error(`Restricted keyword used: ${idStr}`);
        }
    
        // Token identification
        switch (idStr) {
            case 'delvar':
                return { value: idStr, type: TokenType.DeleteVar };
            case 'delfunc':
                return { value: idStr, type: TokenType.DeleteFunc };
            case 'func':
                return { value: idStr, type: TokenType.Func };
            case 'return':
                return { value: idStr, type: TokenType.Return };
            case 'if':
                return { value: idStr, type: TokenType.If };
            case 'else':
                return { value: idStr, type: TokenType.Else };
            case 'let':
                return { value: idStr, type: TokenType.Let };
            case 'make':
                return { value: idStr, type: TokenType.Make };
            case 'show':
                return { value: idStr, type: TokenType.Show };
            case 'error':
                return { value: idStr, type: TokenType.Error };
            case 'alert':
                return { value: idStr, type: TokenType.Alert };
            case 'loop':
                return { value: idStr, type: TokenType.Loop };
            case 'while':
                return { value: idStr, type: TokenType.While };
            case 'run':
                return { value: idStr, type: TokenType.Run };
            case 'connect':
                return { value: idStr, type: TokenType.Connect };
            case 'async':
                return { value: idStr, type: TokenType.Async };
            case 'arradd':
                return { value: idStr, type: TokenType.ArrayAdd };
            case 'arrsize':
                return { value: idStr, type: TokenType.ArrayLength };
            case 'arrdel':
                return { value: idStr, type: TokenType.ArrayRemove };
            case 'read':
                return { value: idStr, type: TokenType.Read };
            case 'arr':
                return { value: idStr, type: TokenType.Array };
            case 'tojson':
                return { value: idStr, type: TokenType.ToJSON };
            case 'parjson':
                return { value: idStr, type: TokenType.ParseJSON };
            case 'return':
                return { value: idStr, type: TokenType.Return }
            case 'call':
                return { value: idStr, type: TokenType.Call }
            case 'floor':
                return { value: idStr, type: TokenType.Floor }
            case 'round':
                return { value: idStr, type: TokenType.Round }
            case 'sqrt':
                return { value: idStr, type: TokenType.SquareRoot }
            case 'sin':
                return { value: idStr, type: TokenType.Sin }
            case 'cos':
                return { value: idStr, type: TokenType.Cos }
            case 'tan':
                return { value: idStr, type: TokenType.Tan }
            case 'tokenize':
                return { value: idStr, type: TokenType.Tokenize }
            case 'lowercase':
                return { value: idStr, type: TokenType.ToLowerCase }
            case 'uppercase':
                return { value: idStr, type: TokenType.ToUpperCase }
            case 'reverse':
                return { value: idStr, type: TokenType.ReverseFunc }
            default:
                return { value: idStr, type: TokenType.Identifier };
        }
    }

    readNumber() {
        let numStr = '';
        let isDecimal = false; // Flag to track if we've encountered a decimal point
    
        while (this.pos < this.input.length) {
            const currentChar = this.input[this.pos];
    
            if (/\d/.test(currentChar)) {
                numStr += currentChar; // Add digit to numStr
                this.pos++;
            } else if (currentChar === '.' && !isDecimal) {
                isDecimal = true; // Mark that we have seen a decimal point
                numStr += currentChar; // Add decimal point to numStr
                this.pos++;
            } else {
                break; // Exit the loop if we hit a non-digit, non-decimal character
            }
        }
    
        if (numStr.length === 0) {
            throw new Error('Invalid number format'); // Optional: throw an error if no digits found
        }
    
        return { value: parseFloat(numStr), type: TokenType.Number }; // Return parsed float
    }
    

    readString(quoteType) {
        let str = '';
        this.pos++;
        while (this.pos < this.input.length && this.input[this.pos] !== quoteType) {
            str += this.input[this.pos++];
        }
        if (this.pos < this.input.length) {
            this.pos++;
        } else {
            throw new Error('Unterminated string literal');
        }
        return { value: str, type: TokenType.String };
    }

    readBinaryOperator(currentChar) {
        let operator = currentChar;
        this.pos++;
        return { value: operator, type: TokenType.BinaryOperator };
    }

    readComparisonOperator(currentChar) {
        let operator = currentChar;
        this.pos++;

        // Check for multi-character operators (e.g., '>=', '<=', '==', '!=')
        if (this.pos < this.input.length && this.input[this.pos] === '=') {
            operator += '=';
            this.pos++;
        }

        return { value: operator, type: TokenType.BinaryOperator };
    }
}

export default Lexer;

File: rust interpreter/KorvaqScrip/src/lexer.rs
use std::collections::HashSet;
use crate::token_type::TokenType;

#[derive(Debug, Clone)]
pub struct Token {
    pub value: String,
    pub token_type: TokenType,
}

pub struct Lexer<'a> {
    pos: usize,
    input: &'a str,
    restricted_keywords: HashSet<String>,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        let restricted_keywords = [
            "var", "const", "for", "switch", "case", "break",
            "continue", "default", "class", "extends", "super", "this",
            "typeof", "instanceof", "delete", "new", "in",
            "try", "catch", "finally", "throw", "debugger",
        ].iter().cloned().map(String::from).collect();

        Lexer {
            pos: 0,
            input,
            restricted_keywords,
        }
    }

    pub fn next_token(&mut self) -> Option<Token> {
        while self.pos < self.input.len() {
            let current_char = self.input[self.pos..].chars().next().unwrap();

            // Skip whitespace
            if current_char.is_whitespace() {
                self.pos += 1;
                continue;
            }

            // Check for single-line comments
            if current_char == '/' && self.input.get(self.pos + 1..).map_or(false, |s| s.starts_with('/')) {
                while self.pos < self.input.len() && self.input.chars().nth(self.pos).unwrap() != '\n' {
                    self.pos += 1;
                }
                self.pos += 1; // Move past the newline
                continue;
            }

            // Check for boolean literals
            if self.input.get(self.pos..).map_or(false, |s| s.starts_with("true")) {
                self.pos += 4; // Move position past "true"
                return Some(Token { value: "true".to_string(), token_type: TokenType::BooleanLiteral });
            }
            if self.input.get(self.pos..).map_or(false, |s| s.starts_with("false")) {
                self.pos += 5; // Move position past "false"
                return Some(Token { value: "false".to_string(), token_type: TokenType::BooleanLiteral });
            }

            // Check for numbers, strings, keywords, operators, and punctuation
            if current_char.is_digit(10) {
                return Some(self.read_number());
            }
            if current_char == '"' || current_char == '\'' || current_char == '`' {
                return Some(self.read_string(current_char));
            }
            if current_char.is_alphabetic() || current_char == '_' {
                return Some(self.read_keyword_or_identifier());
            }

            // Handle operators and punctuation
            if current_char == '=' {
                self.pos += 1;
                if self.pos < self.input.len() && self.input.chars().nth(self.pos).unwrap() == '=' {
                    self.pos += 1;
                    if self.pos < self.input.len() && self.input.chars().nth(self.pos).unwrap() == '=' {
                        self.pos += 1;
                        return Some(Token { value: "===" .to_string(), token_type: TokenType::Equals });
                    }
                    return Some(Token { value: "==" .to_string(), token_type: TokenType::Equals });
                }
                return Some(Token { value: "=" .to_string(), token_type: TokenType::Equals });
            }
            if current_char == '(' {
                self.pos += 1;
                return Some(Token { value: "(".to_string(), token_type: TokenType::OpenParen });
            }
            if current_char == ')' {
                self.pos += 1;
                return Some(Token { value: ")".to_string(), token_type: TokenType::CloseParen });
            }
            // Ignore semicolon
            if current_char == ';' {
                self.pos += 1;
                continue; // Skip semicolon
            }
            if current_char == ':' {
                self.pos += 1;
                return Some(Token { value: ":".to_string(), token_type: TokenType::Colon });
            }
            if current_char == '{' {
                self.pos += 1;
                return Some(Token { value: "{".to_string(), token_type: TokenType::OpenBrace });
            }
            if current_char == '}' {
                self.pos += 1;
                return Some(Token { value: "}".to_string(), token_type: TokenType::CloseBrace });
            }

            // Handle array syntax
            if current_char == '[' {
                self.pos += 1;
                return Some(Token { value: "[".to_string(), token_type: TokenType::OpenBracket });
            }
            if current_char == ']' {
                self.pos += 1;
                return Some(Token { value: "]".to_string(), token_type: TokenType::CloseBracket });
            }
            if current_char == ',' {
                self.pos += 1;
                return Some(Token { value: ",".to_string(), token_type: TokenType::Comma });
            }

            if current_char == '*' && self.input.get(self.pos + 1..).map_or(false, |s| s.starts_with("*")) {
                self.pos += 2; // Move past '**'
                return Some(Token { value: "**".to_string(), token_type: TokenType::BinaryOperator });
            }

            // Handle binary operators
            if ['+', '-', '*', '/', '%', '^'].contains(&current_char) {
                return Some(self.read_binary_operator(current_char));
            }

            // Handle comparison operators
            if current_char == '>' {
                return Some(self.read_comparison_operator('>'));
            }
            if current_char == '<' {
                return Some(self.read_comparison_operator('<'));
            }

            // Handle logical operators
            if current_char == '&' && self.input.get(self.pos + 1..).map_or(false, |s| s.starts_with("&")) {
                self.pos += 2; // Move past '&&'
                return Some(Token { value: "&&".to_string(), token_type: TokenType::LogicalAnd });
            }
            if current_char == '|' && self.input.get(self.pos + 1..).map_or(false, |s| s.starts_with("|")) {
                self.pos += 2; // Move past '||'
                return Some(Token { value: "||".to_string(), token_type: TokenType::LogicalOr });
            }

            // Handle not equal operator
            if current_char == '!' && self.input.get(self.pos + 1..).map_or(false, |s| s.starts_with("=")) {
                self.pos += 2; // Move past '!='
                return Some(Token { value: "!=".to_string(), token_type: TokenType::NotEquals });
            }

            panic!("Unexpected character: {}", current_char);
        }

        None
    }

    fn read_keyword_or_identifier(&mut self) -> Token {
        let mut id_str = String::new();

        // Allow the first character to be a letter or underscore
        if self.input[self.pos..].chars().next().unwrap().is_alphabetic() || self.input[self.pos..].chars().next().unwrap() == '_' {
            id_str.push(self.input[self.pos..].chars().next().unwrap());
            self.pos += 1;
        } else {
            panic!("Invalid identifier start: {} at position {}", self.input[self.pos..].chars().next().unwrap(), self.pos);
        }

        // Allow subsequent characters to be letters, digits, or underscores
        while self.pos < self.input.len() && (self.input[self.pos..].chars().next().unwrap().is_alphanumeric() || self.input[self.pos..].chars().next().unwrap() == '_') {
            id_str.push(self.input[self.pos..].chars().next().unwrap());
            self.pos += 1;
        }

        // Check for restricted keywords
        if self.restricted_keywords.contains(&id_str) && id_str != "let" {
            panic!("Using a restricted keyword: {}", id_str);
        }

        match id_str.as_str() {
            "let" => Token { value: id_str, token_type: TokenType::Let },
            "make" => Token { value: id_str, token_type: TokenType::Make },
            "show" => Token { value: id_str, token_type: TokenType::Show }, 
            "delvar" => Token { value: id_str, token_type: TokenType::DelVar },
            "if" => Token { value: id_str, token_type: TokenType::If },
            "else" => Token { value: id_str, token_type: TokenType::Else },
            "uppercase" => Token { value: id_str, token_type: TokenType::UpperCase },
            "lowercase" => Token { value: id_str, token_type: TokenType::LowerCase },
            _ => Token { value: id_str, token_type: TokenType::Identifier },
        }
    }

    fn read_number(&mut self) -> Token {
        let start_pos = self.pos;
        while self.pos < self.input.len() && self.input[self.pos..].chars().next().unwrap().is_digit(10) {
            self.pos += 1;
        }

        let number_str = &self.input[start_pos..self.pos];
        Token { value: number_str.to_string(), token_type: TokenType::Number }
    }

    fn read_string(&mut self, quote_char: char) -> Token {
        self.pos += 1; // Skip the opening quote
        let start_pos = self.pos;

        while self.pos < self.input.len() && self.input[self.pos..].chars().next().unwrap() != quote_char {
            self.pos += 1;
        }

        let string_content = &self.input[start_pos..self.pos];
        self.pos += 1; // Skip the closing quote
        Token { value: string_content.to_string(), token_type: TokenType::String }
    }

    fn read_binary_operator(&mut self, op: char) -> Token {
        self.pos += 1; // Move past operator
        Token { value: op.to_string(), token_type: TokenType::BinaryOperator }
    }

    fn read_comparison_operator(&mut self, op: char) -> Token {
        self.pos += 1; // Move past operator
        Token { value: op.to_string(), token_type: TokenType::BinaryOperator }
    }
}
